File: frontend/src/types/execution.ts
export type ExecutionStatus = "PENDING" | "RUNNING" | "SUCCESS" | "ERROR" | "CANCELLED" | "SKIPPED";

export type TriggerType = "MANUAL" | "SCHEDULE" | "POLLING" | "AGENT";

export interface ExecutionStep {
  id: number;
  execution_id: number;
  node_id: string;
  node_type: string;
  status: ExecutionStatus;
  started_at: string;
  finished_at: string | null;
  input_data: Record<string, unknown> | null;
  output_data: Record<string, unknown> | null;
  error: string | null;
  duration_ms: number | null;
}

export interface Execution {
  id: number;
  workflow_id: number;
  status: ExecutionStatus;
  trigger_type: TriggerType;
  started_at: string;
  finished_at: string | null;
  input_data: Record<string, unknown> | null;
  output_data: Record<string, unknown> | null;
  error: string | null;
}

export interface ExecutionDetail extends Execution {
  steps: ExecutionStep[];
}

File: backend/app/schemas/execution.py
"""Pydantic schemas for Execution API."""
from datetime import datetime
from typing import Any
from pydantic import BaseModel, ConfigDict

from app.models.execution import ExecutionStatus, TriggerType


class ExecutionStepOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    node_id: str
    node_type: str
    status: ExecutionStatus
    started_at: datetime
    finished_at: datetime | None
    input_data: dict[str, Any]
    output_data: dict[str, Any]
    error: str
    duration_ms: int


class ExecutionOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    workflow_id: int
    status: ExecutionStatus
    trigger_type: TriggerType
    started_at: datetime | None
    finished_at: datetime | None
    input_data: dict[str, Any]
    output_data: dict[str, Any]
    error: str
    created_at: datetime


class ExecutionDetailOut(ExecutionOut):
    steps: list[ExecutionStepOut]
